### PR TITLE
Adding custom menu options when making a menu

### PR DESCRIPTION
**Pull request by mertasan**
# Edit:

**settings content:**
settings

**example 1:**

**example 1 results:**

1results

**example 2:**

**example 2 results:**

screenshot 31

**example 3:**

**example 3 results:**

3results

**example 4:**

**example 4 results:**

4results

### Edit 2:

> Results are printed from the view file. Menus were created with middleware.

screenshot 33

screenshot 32

### Edit 3:

You can override the default settings with the following methods. Or you can add new custom settings to the menu.

```php
return [
    'default' => [
        'auto_activate'    => true,
        'activate_parents' => true,
        'active_class'     => 'active',
        'active_element'   => 'item',    // item|link
        'restful'          => true,
    ],
    'mysidebar' => [
        'active_class'     => 'active-class-mysidebar',
    ],
    'mynavbar' => [
        'active_class'     => 'active-class-mynavbar',
    ],
];
```

**Example 1:**

Override default settings

```php
Menu::make('MySidebar', function ($menu) {

    $menu->options([
        'active_class' => 'new-active-class',
    ]);

    $menu->add('Home');
    $menu->add('About', 'about');
    
});

/**
 * Results:
[
    'auto_activate'    => true,
    'activate_parents' => true,
    'active_class'     => 'new-active-class'
    'active_element'   => 'item',
    'restful'          => true,
]
*/
```

**Example 2:**

Add new settings specific to the menu.

```php
Menu::make('MySidebar', function ($menu) {

    $menu->options([
        'inactive_class' => 'custom-inactive-class-mysidebar',
    ]);

    $menu->add('Home');
    $menu->add('About', 'about');
    
});

/**
 * Results:
[
    'auto_activate'    => true,
    'activate_parents' => true,
    'active_class'     => 'active-class-mysidebar'
    'active_element'   => 'item',
    'restful'          => true,
    'inactive_class'   => 'custom-inactive-class-mysidebar',
]
*/
```

**Example 3:**

Add custom settings and get the rest of the settings from `MySidebar`.

```php
Menu::make('MyNavbar', function ($menu) {

    $menu->options([
        'inactive_class' => 'custom-inactive-class-mynavbar',
    ], 'MySidebar'); // or mysidebar

    $menu->add('Home');
    $menu->add('About', 'about');
    
});

/**
 * Results:
[
    'auto_activate'    => true,
    'activate_parents' => true,
    'active_class'     => 'active-class-mysidebar'
    'active_element'   => 'item',
    'restful'          => true,
    'inactive_class'   => 'custom-inactive-class-mynavbar',
]
*/
```

**Example 4:**

Override all settings (including default settings) and add new ones.

```php
Menu::make('MyNavbar', function ($menu) {

    $menu->options([
        'active_class' => 'active',
    ], null); 

    $menu->add('Home');
    $menu->add('About', 'about');
    
});

/**
 * Results:
[
    'active_class'     => 'active'
]
*/
```

**Example 5:**

Or you can use it like this

```php
Menu::make('MyNavbar', function ($menu) {

    $menu->add('Home');
    $menu->add('About', 'about');
    
}, [
    'inactive_class' => 'custom-inactive-class-mynavbar',
]);

/**
 * Results:
[
    'auto_activate'    => true,
    'activate_parents' => true,
    'active_class'     => 'active-class-mynavbar'
    'active_element'   => 'item',
    'restful'          => true,
    'inactive_class'   => 'custom-inactive-class-mynavbar',
]
*/
```

### CODE REVIEW

File: src/Lavary/Menu/Builder.php [hunk 1]
```diff
@@ -758,6 +758,24 @@ public function conf($key)
...
+     *
+     * @param array $config
+     * @param bool $merge
+     * @param string $configGroup
+     *
+     * @return string
+     */
+    public function config(array $config, bool $merge = true, string $configGroup = 'default')
+    {
+        if($merge){
```
Comment (dustingraham) [inline src/Lavary/Menu/Builder.php:772]: `if($merge){` -- Could you please follow the syntax of nearby code. So, to match the function above, this should be `if ($merge) {`

File: src/Lavary/Menu/Builder.php [hunk 2]
```diff
@@ -758,6 +758,24 @@ public function conf($key)
...
+            $this->conf = array_replace_recursive(config('laravel-menu.settings.'.$configGroup, []), $config);
```
Comment (dustingraham) [inline src/Lavary/Menu/Builder.php:773]: I think this is a little confusing. At first it looks like a recursive call to $this->config(...) where the signature doesn't match. But, instead this is a call to laravel's config(...) method. What is it doing (can you add a comment to the code?) is it saving this config globally? What's the point, does it allow easy reuse in another method?

**Comment by dustingraham** [comment]
This looks awesome, and incredibly handy. I love convenience functions like this.

The examples you put in this PR should be added to the main readme. Plus, I'm not sure I fully understand what configGroup and the config(..., []) call is doing here yet.

**Comment by mertasan** [comment]
@dustingraham glad you liked it. I will update the PR. (and i will add in new examples.) 

also, I will prepare another PR for using icons. 

ready in a few hours.

**Comment by mertasan** [comment]
@dustingraham Can you review the latest changes? Also I edited the original post.

**Comment by mertasan** [comment]
I have added new examples to my original post. (**_Edit 2_**)

**Comment by dustingraham** [comment]
@mertasan thank you for the screenshots, but it is much better if you use the code highlighting so that users viewing this pull request can copy/paste the text.

Example code formatting. (This is what you had originally, and it is much easier to use than the screenshots.)

```
  // Add & Overwrite "yourmenuname" config group
  \Menu::make('MyNavBar', function ($menu) {

      $menu->config([
          'active_class' => 'custom-class'
      ], true, 'yourmenuname');

      $menu->add('Home');
      $menu->add('About', 'about');
  });
```

Also, my request was that you update https://github.com/lavary/laravel-menu/blob/master/README.md with your examples. These examples like the one I just copied are good to be added to the readme in a section after the main config options sections.

**Comment by mertasan** [comment]
@dustingraham my goal was to show the type hints generated by the IDE. ok i am adding. thank you

**Comment by dustingraham** [comment]
I use jetbrains products as well, and I love the type hints, but I don't think that's necessary. I agree and understand the benefits of what you're doing.

What users will find most commonly useful are going to be the three examples you provided. Those are great functional examples of how the config function can be used.

**Comment by mertasan** [comment]
My goal was to show what I was doing. For the next step, I planned to add code blocks instead of screenshots. I agree with what you said. I am very glad you liked it. thank you.

I have added new examples to my original post. (_**Edit 3**_)

**Comment by dustingraham** [comment]
I've looked through this, and it looks good. I need to download it and test it, then I will merge. Thank you for the great documentation additions.

**Comment by mertasan** [comment]
@dustingraham I am using these changes in my project. I haven't had much time to wait. That's why I opened a new warehouse. It was a well-meaning thing. I wanted you to know. https://github.com/mertasan/laravel-menu#thanks-to

**Comment by dustingraham** [comment]
> @dustingraham I am using these changes in my project. I haven't had much time to wait. That's why I opened a new warehouse. It was a well-meaning thing. I wanted you to know. https://github.com/mertasan/laravel-menu#thanks-to

Completely understandable. I just need to have some time available to test before merging and publishing a new version on this package since I don't want to break anyone else's implementation.

I appreciate both of these PR's and I intend to merge them.

**Comment by dustingraham** [comment]
Sorry for the delay. I had to set up a whole new local project to test with. Tested and this looks great.

**Pull request merged**